server/src/main/java/org/elasticsearch/transport/TransportRequestDeduplicator.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Deduplicator for {@link TransportRequest}s that keeps track of {@link TransportRequest}s that should
+ * not be sent in parallel.
+ * @param <T> Transport Request Class
+ */
+public final class TransportRequestDeduplicator<T extends TransportRequest> {
+
+    private final ConcurrentMap<T, CompositeListener> requests = ConcurrentCollections.newConcurrentMap();
+
+    /**
+     * Register a listener for the given request with the deduplicator.
+     * If the request is not yet registered with the deduplicator it will return an {@link ActionListener}
+     * that must be completed by the called when the request completes. If the request is already known to
+     * the deduplicator it will keep track of the given listener and invoke it when the listener returned
+     * for the first invocation with the request is completed.
+     * The caller of this method should therefore execute the transport request if returned an instance
+     * of {@link ActionListener} and invoke that listener when completing the request and do nothing when
+     * being returned {@code null}.
+     * @param request Request to deduplicate
+     * @param listener Listener to invoke on request completion
+     * @return Listener that must be invoked by the caller or null when the request is already known
+     */
+    public ActionListener<Void> register(T request, ActionListener<Void> listener) {
```

Comment:
I am making a suggestion but feel free to reject it because for me this PR is very good already. How about adding a new argument  `Consumer<Listener>` that executes the request if it was not executed before (surely we need to rename the method). With this change, I think we can safely use this method without reading the documentation.

Reply:
I think we miss a test which ensures that if we register a request multiple times (maybe concurrently), we return a non-null result only once. And we need also to verify that when the returned listener is completed, the cached size is reduced.

Reply:
> How about adding a new argument Consumer<Listener> that executes the request if it was not executed before (surely we need to rename the method). With this change, I think we can safely use this method without reading the documentation.

👍 I like it :) On it.

> I think we miss a test which ensures that if we register a request multiple times (maybe concurrently), we return a non-null result only once. And we need also to verify that when the returned listener is completed, the cached size is reduced.

Sure, sounds reasonable => on it :)